git.tex

```diff
@@ -1,11 +1,38 @@
+[TODO: describe in terms of Agile values?]
 
-Feature branches associated with Jira tickets.
+\subsubsection{Version Control}\label{sec:git}\label{sec:subversion}
 
-Branch protection.
+[TODO: why did LSST switch from Subversion to Git?]
 
-Rebase before merging.
+\subsubsection{Code Organization}\label{sec:git_repositories}
+
+The data management system is divided into several hundred individual Git repositories, hosted on GitHub\footnote{\url{http://github.com/}}.
+Each repository corresponds to one code package, and
+all but a handful of repositories usually have at most one developer working on them at a time.
+Isolating code in small repositories minimizes the need for developers to frequently download code updates, something the Git framework requires whenever developers make concurrent changes to the \texttt{master} branch of the same repository.
+Representing each package by an independent repository also makes it easy to formalize and track inter-package dependencies using tools like EUPS (\autoref{sec:eups}).
+
+\subsubsection{The LSST Workflow}\label{sec:dev_workflow}
+
+LSST software development is based on the shared repository model:
+work is done on branches of a single repository rather than on user forks.
+Development work is associated with a Jira ticket (\autoref{sec:jira_ticket}), and
+keeping all work in a shared repository makes it easier to link code changes to their corresponding Jira tickets.
+[TODO: why is that important? I assume it has something to do with earned value tracking]
+
+All work associated with a specific Jira ticket is done on a feature branch named after that ticket.
+Developers must test their branch as part of the full data management system (\autoref{sec:jenkins}) before the code is considered ready to review or merge.
+GitHub pull requests from the ticket branch to \texttt{master} are used for code review, but not to merge the code.
+Instead, feature branches are merged by first rebasing them onto the latest version of \texttt{master}, then manually merging them from the command line.
+This approach ensures the Git commit history appears linear to later inspection, without the need for possibly complicated merge commits.
+The merging process does create an empty merge commit that serves as a marker to distinguish which commits on \texttt{master} came from which feature branch.
+
+Originally, the workflow described above was enforced by individual developers and their reviewers.
+While it worked well most of the time, there were occasional cases where a developer would accidentally push commits directly to \texttt{master}.
+Therefore, we began using GitHub's branch protection feature to block both direct commits to \texttt{master} and merges of unrebased ticket branches.
+Branch protection is transparent to developers so long as they follow the workflow, so this improvement catches errors without altering the overall development process.
```

Comment:
We have historically allowed direct commits to master of documentation fixes but now that doc fixes can cause flake8 failures we no longer allow this.

Reply:
Are you asking for me to include an aside about documentation as opposed to code?

Reply:
I'm talking about fixes to docstrings in code, so I think it's worth mentioning that we used to allow them direct to master but now require PRs because they can break tests.